pkg/controller/operators/labeller/labels.go

```diff
@@ -0,0 +1,153 @@
+package labeller
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	jsonpatch "github.com/evanphx/json-patch"
+	"github.com/sirupsen/logrus"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+
+	operatorsv1 "github.com/operator-framework/api/pkg/operators/v1"
+	operatorsv1alpha1 "github.com/operator-framework/api/pkg/operators/v1alpha1"
+	operatorsv1alpha2 "github.com/operator-framework/api/pkg/operators/v1alpha2"
+
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/controller/install"
+	"github.com/operator-framework/operator-lifecycle-manager/pkg/controller/operators/decorators"
+)
+
+type ApplyConfig[T any] interface {
+	WithLabels(map[string]string) T
+}
+
+type Client[A ApplyConfig[A], T metav1.Object] interface {
+	Apply(ctx context.Context, cfg ApplyConfig[A], opts metav1.ApplyOptions) (result T, err error)
+}
+
+func hasLabel(obj metav1.Object) bool {
+	if labels := obj.GetLabels(); labels != nil {
+		if value, ok := labels[install.OLMManagedLabelKey]; ok && value == install.OLMManagedLabelValue {
+			return true
+		}
+	}
+	return false
+}
+
+func ObjectLabeler[T metav1.Object, A ApplyConfig[A]](
+	ctx context.Context,
+	logger *logrus.Logger,
+	check func(metav1.Object) bool,
+	applyConfigFor func(name, namespace string) A,
+	apply func(namespace string, ctx context.Context, cfg A, opts metav1.ApplyOptions) (T, error),
+) func(
+	obj interface{},
+) error {
```

Comment:
Maybe, for readability, give this a custom type?

Reply:
Sorry, a bit hard to trace what you're commenting on here - the `func(obj interface{}) error` ?

Reply:
Yes, the return